### PR TITLE
feat(saml): add SSO authorization API (disabled by default)

### PR DESCRIPTION
**Pull request by hf**
Adds the SSO authorization API with which you can initiate a single sign-on flow. 

You can do this by sending a POST request to `/sso`. The parameters are as follows:

```json
{
  "provider_id": "cb059692-59ec-4ac3-848a-0d6b48569eac"
}
```

Which will initiate SSO with a known provider ID.

```json
{
  "domain": "example.com"
}
```

Which will initiate SSO by looking up the registered domain for a provider. This is the more useful version of the call.

The call results in a [303 See Other](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/303) response that redirects to the provider's authorization URL.

[PR for gotrue-js.](https://github.com/supabase/gotrue-js/pull/526)

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -156,6 +156,8 @@ func NewAPIWithVersion(ctx context.Context, globalConfig *conf.GlobalConfigurati
 
 		if api.config.SAML.Enabled {
 			r.Route("/sso", func(r *router) {
+				r.Post("/", api.SingleSignOn)
```
Comment (hf) [inline api/api.go:159]: This needs a rate limited and hCaptcha.

File: api/sso_test.go [hunk 2]
```diff
@@ -515,6 +516,128 @@ func (ts *SSOTestSuite) TestAdminDeleteSSOProvider() {
...
+		providers[i].ID = payload.ID
+	}
+
+	examples := []struct {
+		Code    int
+		Request map[string]interface{}
+		URL     string
+	}{
+		{
+			Request: map[string]interface{}{
```
Comment (J0) [inline api/sso_test.go:569]: /nit would be nice to have test case descriptions

**Review by J0 — APPROVED** [review]
LGTM

File: api/singlesignon.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
+package api
```
Comment (kangmingtay) [inline api/singlesignon.go:1]: nit: rename file name to `sso.go`?

File: api/singlesignon.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+			return notFoundError("No SSO provider assigned for this domain")
+		} else if err != nil {
+			return internalServerError("Unable to find SSO provider by domain").WithInternalError(err)
+		}
+	}
+
+	if ssoProvider == nil {
+		// impossible condition since #validate() should have handled this case
+		return internalServerError("Unable to find SSO provider")
+	}
```
Comment (kangmingtay) [inline api/singlesignon.go:71]: nit: might want to consider making `validate()` return `(hasProviderId bool, err error)` so you don't have to check `params.ProviderID != uuid.Nil` and `params.Domain != ""` again

Comment (kangmingtay) [reply api/singlesignon.go:71]: and you can do away with the LOC with the impossible condition clause

File: api/sso_test.go (hunk 2 above)
Comment (kangmingtay) [reply api/sso_test.go:569]: +1 on this - would make the test cases easier to understand!

Comment (kangmingtay) [reply api/sso_test.go:569]: +1 on this - would make the test cases easier to understand!

**Review by kangmingtay — APPROVED** [review]

**Comment by kangmingtay** [comment]
@hf just wanted to check my understanding here, `POST /sso` is similar to the `GET /authorize` step in the oauth flow right? i was wondering where a developer would generate the `provider_id` to pass into the request body? is that done in a separate API call somewhere else?

File: api/singlesignon.go (hunk 3 above)
Comment (hf) [reply api/singlesignon.go:1]: There's already a `sso.go` file that hosts the admin methods. But maybe I'll rename that to `ssoadmin.go` and this to `sso.go`. Great catch!

**Comment by hf** [comment]
> @hf just wanted to check my understanding here, `POST /sso` is similar to the `GET /authorize` step in the oauth flow right? i was wondering where a developer would generate the `provider_id` to pass into the request body? is that done in a separate API call somewhere else?

Yes, you can get a provider's ID from the admin list providers call typically. This is useful in cases where you would have a provider that does not have a domain associated (honestly, this basically never happens IRL). Could be useful in integration testing scenarios and the like.

**Pull request merged**